src/Symfony/Bundle/WebProfilerBundle/Tests/Controller/ProfilerControllerTest.php

```diff
@@ -11,19 +11,110 @@
 
-class ProfilerControllerTest extends TestCase
+class ProfilerControllerTest extends WebTestCase
 {
+    public function testHomeActionWithProfilerDisabled()
+    {
+        $this->expectException(NotFoundHttpException::class);
+        $this->expectExceptionMessage('The profiler must be enabled.');
+
+        $urlGenerator = $this->getMockBuilder('Symfony\Component\Routing\Generator\UrlGeneratorInterface')->getMock();
+        $twig = $this->getMockBuilder('Twig\Environment')->disableOriginalConstructor()->getMock();
+
+        $controller = new ProfilerController($urlGenerator, null, $twig, []);
+        $controller->homeAction();
+    }
+
+    public function testHomeActionRedirect()
+    {
+        $kernel = new WebProfilerBundleKernel();
+        $client = new Client($kernel);
+
+        $client->request('GET', '/_profiler/');
+
+        $this->assertSame(302, $client->getResponse()->getStatusCode());
+        $this->assertSame('/_profiler/empty/search/results?limit=10', $client->getResponse()->getTargetUrl());
+    }
+
+    public function testPanelActionWithLatestTokenWhenNoTokensExist()
+    {
+        $kernel = new WebProfilerBundleKernel();
+        $client = new Client($kernel);
+
+        $client->request('GET', '/_profiler/latest');
+
+        $this->assertStringContainsString('No profiles found.', $client->getResponse()->getContent());
+    }
+
+    public function testPanelActionWithLatestToken()
+    {
+        $kernel = new WebProfilerBundleKernel();
+        $client = new Client($kernel);
+
+        $client->request('GET', '/');
+        $client->request('GET', '/_profiler/latest');
+
+        $this->assertStringContainsString('kernel:homepageController', $client->getResponse()->getContent());
+    }
+
+    public function testPanelActionWithoutValidToken()
+    {
+        $kernel = new WebProfilerBundleKernel();
+        $client = new Client($kernel);
+
+        $client->request('GET', '/_profiler/this-token-does-not-exist');
+
+        $this->assertStringContainsString('Token &quot;this-token-does-not-exist&quot; not found.', $client->getResponse()->getContent());
+    }
+
+    public function testPanelActionWithWrongPanel()
+    {
+        $kernel = new WebProfilerBundleKernel();
+        $client = new Client($kernel);
+
+        $client->request('GET', '/');
+        $client->request('GET', '/_profiler/latest?panel=this-panel-does-not-exist');
+
+        $this->assertSame(404, $client->getResponse()->getStatusCode());
+    }
+
+    public function testPanelActionWithValidPanelAndToken()
+    {
+        $kernel = new WebProfilerBundleKernel();
+        $client = new Client($kernel);
+
+        $client->request('GET', '/');
+        $crawler = $client->request('GET', '/_profiler/latest?panel=router');
+
+        $this->assertSame('_', $crawler->filter('.metrics .metric .value')->eq(0)->text());
+        $this->assertSame('12', $crawler->filter('.metrics .metric .value')->eq(1)->text());
```

Comment:
I removed this assertion when merging in 4.4 because it doesn't pass and the patc is painful enough to merge up.
Feel free to submit it again on 4.4 with the proper assertion.